src/x86/nt.rs

```diff
@@ -0,0 +1,414 @@
+    )]
+    ///
+    /// Note that the memory must be borrowed for the *whole* duration of the original scope.
+    /// Automatic reborrowing can not be used to shorten it for a different scope.
+    ///
+    /// ```rust,compile_fail
+    #[cfg_attr(
+        target_arch = "x86",
+        doc = "\
+            use safe_unaligned_simd::x86::NonTemporalScope;\n\
+            use core::arch::x86::__m256i;\n\
+        "
+    )]
+    #[cfg_attr(
+        target_arch = "x86_64",
+        doc = "\
+            use safe_unaligned_simd::x86_64::NonTemporalScope;\n\
+            use core::arch::x86_64::__m256i;\n\
+        "
+    )]
+    ///
+    /// #[target_feature(enable = "avx")]
+    /// fn zero_data<'d>(scope: NonTemporalScope<'d>, data: &'d mut __m256i) {
+    ///     let first = scope.prepare_write(data);
+    ///     // Fails!
+    ///     let second = scope.prepare_write(data);
+    /// }
+    /// ```
+    pub fn prepare_write<T>(&self, inner: &'data mut T) -> NonTemporalStoreable<'data, T> {
+        NonTemporalStoreable {
+            inner: ptr::NonNull::from(inner),
+            marker: PhantomData,
+        }
+    }
+
+    /// Run a closure with the guarantee of exiting with an `sfence` instruction. A closure is
+    /// invoked within the scope and given a value in reference to the scope. That value allows
+    /// qualifying mutably owned memory as memory which can be targeted with non-temporal stores.
+    /// When memory is wrapped in such a manner, no other access is allowed until the scope exits.
+    ///
+    /// ```
+    #[cfg_attr(
+        target_arch = "x86",
+        doc = "\
+            use safe_unaligned_simd::x86::{NonTemporalScope, _mm256_stream_si256};\n\
+            use core::arch::x86::{__m256i, _mm256_set1_epi8};\n\
+        "
+    )]
+    #[cfg_attr(
+        target_arch = "x86_64",
+        doc = "\
+            use safe_unaligned_simd::x86_64::{NonTemporalScope, _mm256_stream_si256};\n\
+            use core::arch::x86_64::{__m256i, _mm256_set1_epi8};\n\
+        "
+    )]
+    /// #[target_feature(enable = "avx")]
+    /// fn zero_data<'d>(scope: NonTemporalScope<'d>, data: &'d mut [__m256i]) {
+    ///     let v = _mm256_set1_epi8(0);
+    ///
+    ///     for part in data {
+    ///         let mut storeable = scope.prepare_write(part);
+    ///         _mm256_stream_si256(&mut storeable, v);
+    ///     }
+    /// }
+    ///
+    /// # #[target_feature(enable = "avx")]
+    /// # fn _do_main() {
+    /// let data: &mut [__m256i] = // ..
+    /// # &mut [_mm256_set1_epi8(1); 4];
+    ///
+    /// NonTemporalScope::with(|scope| {
+    ///     zero_data(scope, data);
+    /// });
+    /// # let a: [u16; 16] = unsafe { core::mem::transmute(data[0]) };
+    /// # assert_eq!(a, [0; 16]);
+    /// # }
+    /// #
+    /// # // Miri does not support non-temporal instructions, avx needed for the example
+    /// # if cfg!(all(target_feature = "avx", miri)) {
+    /// #     unsafe { _do_main() }
+    /// # }
+    /// ```
```

Comment:
Question: How does this pass Miri?

I would have expected this to be a `not(miri)` sort of cfg since Miri doesn't support the streaming stores.

Reply:
Miri does not execute the code. It still compiles everything but that `cfg` converts to a condition-false at runtime, similar to `ignore` but since I was gating `avx` already it felt a little more natural to do than the intrusive `cfg_attr`.